File: packages/typescript/src/transport/binary.ts
// packages/typescript/src/transport/binary.ts
import { AEPTransportError } from "../errors.js";
import { ARTIFACT_BYTES_PATH, HEADER_PROTOCOL_VERSION, HEADER_SESSION_ID, PROTOCOL_VERSION } from "../constants.js";

export interface BinaryResult {
  mimeType: string;
  size: number;
  body: ReadableStream<Uint8Array>;
}

export class BinaryClient {
  private readonly baseUrl: string;
  private readonly bearer: string;
  private readonly fetchImpl: typeof fetch;

  constructor(config: { baseUrl: string; bearer: string; fetch?: typeof fetch }) {
    this.baseUrl = config.baseUrl.replace(/\/$/, "");
    this.bearer = config.bearer;
    this.fetchImpl = config.fetch ?? fetch;
  }

  async fetchArtifactBytes(
    artifactId: string,
    sessionId: string | null,
    signal?: AbortSignal,
  ): Promise<BinaryResult> {
    const headers: Record<string, string> = {
      "Authorization": `Bearer ${this.bearer}`,
      [HEADER_PROTOCOL_VERSION]: PROTOCOL_VERSION,
    };
    if (sessionId) headers[HEADER_SESSION_ID] = sessionId;

    const resp = await this.fetchImpl(`${this.baseUrl}${ARTIFACT_BYTES_PATH(artifactId)}`, {
      method: "GET",
      headers,
      ...(signal ? { signal } : {}),
    });

    if (!resp.ok) {
      throw new AEPTransportError(`Binary HTTP ${resp.status} ${resp.statusText}`);
    }
    if (!resp.body) {
      throw new AEPTransportError("Binary response has no body");
    }

    return {
      mimeType: resp.headers.get("Content-Type") ?? "application/octet-stream",
      size: Number(resp.headers.get("Content-Length") ?? 0),
      body: resp.body,
    };
  }
}

File: packages/python/helix_sdk/transport/binary.py
# packages/python/helix_sdk/transport/binary.py
from __future__ import annotations

from dataclasses import dataclass
from typing import AsyncIterator

import httpx

from ..constants import HEADER_PROTOCOL_VERSION, HEADER_SESSION_ID, PROTOCOL_VERSION, artifact_bytes_path
from ..errors import AEPTransportError


@dataclass
class BinaryResult:
    mime_type: str
    size: int
    body: AsyncIterator[bytes]


class BinaryClient:
    def __init__(self, base_url: str, bearer: str, http_client: httpx.AsyncClient | None = None) -> None:
        self._base_url = base_url.rstrip("/")
        self._bearer = bearer
        self._client = http_client or httpx.AsyncClient(timeout=httpx.Timeout(60.0, read=None))
        self._owns_client = http_client is None

    async def aclose(self) -> None:
        if self._owns_client:
            await self._client.aclose()

    async def fetch_artifact_bytes(self, artifact_id: str, session_id: str | None) -> BinaryResult:
        headers = {
            "Authorization": f"Bearer {self._bearer}",
            HEADER_PROTOCOL_VERSION: PROTOCOL_VERSION,
        }
        if session_id:
            headers[HEADER_SESSION_ID] = session_id

        req = self._client.build_request(
            "GET",
            f"{self._base_url}{artifact_bytes_path(artifact_id)}",
            headers=headers,
        )
        resp = await self._client.send(req, stream=True)
        if resp.status_code != 200:
            await resp.aclose()
            raise AEPTransportError(f"Binary HTTP {resp.status_code} {resp.reason_phrase}")

        mime = resp.headers.get("Content-Type", "application/octet-stream")
        size = int(resp.headers.get("Content-Length", "0") or 0)

        async def iterator() -> AsyncIterator[bytes]:
            try:
                async for chunk in resp.aiter_bytes():
                    yield chunk
            finally:
                await resp.aclose()

        return BinaryResult(mime_type=mime, size=size, body=iterator())
